### ISSUE TITLE
nolintlint: reports false positive warnings about used nolint directive

### ISSUE THREAD

**Issue by fho** [golangci/golangci-lint#3228]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

We have staticcheck and nonolint enabled.
Sometimes in our CI builds golangci-lint reports a false warnings from nonolint about unused `nolint:staticcheck` comments.
All warnings that I checked are wrong, the `//nolint:staticcheck` directives always silence existing deprecation warnings from staticcheck.

I was not able to reproduce this issue.
In CI we reuse the golangci-lint and go cache when running checks on different branches.
It can happen that the same cache was used previously with older go and/or golangci-lint versions.

I believe https://github.com/golangci/golangci-lint/issues/1940#issuecomment-1239477698 refers to the same issue.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.49.0 built from cc2d97f3 on 2022-08-24T10:24:37Z
```

</details>

### Configuration file

<details>

```console
$ cat .golangci.yml
run:
  concurrency: 2
  deadline: 5m

linters:
  disable-all: true
  enable:
    - depguard
    - errcheck
    - exportloopref
    - goimports
    - goprintffuncname
    - ineffassign
    - misspell
    - nolintlint
    - revive
    - rowserrcheck
    - staticcheck
    - typecheck
    - unconvert
    - unused
    - vet

issues:
  exclude-use-default: false
```

</details>

### Go environment

<details>

```console
$ go version && go env
[/src/go/code/orders]$ go version && go env
go version go1.19.1 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN="/usr/local/bin"
GOCACHE="/var/cache/go-build"
GOENV="/tmp/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS="-tags=timetzdata -trimpath"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/var/cache/go-modcache"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/tmp/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/opt/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/opt/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.19.1"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="0"
GOMOD="/src/go/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1328424340=/tmp/go-build -gno-record-gcc-switches"
```

</details>

### Verbose output of running

N/A

### Code example or link to a public repository

Example case, output from golangci-lint:
```
[..]/itemdetails.go:36:32: directive `//nolint:staticcheck // Description field is deprecated` is unused for linter "staticcheck" (nolintlint)
		return item.Description, nil //nolint:staticcheck // Description field is deprecated
```

The referenced Description field has a deprecation comment set:
```golang
// Deprecated: Do not use.
Description      string                 `protobuf:"bytes,4,opt,name=description,proto3" json:"description,omitempty"`
```

**Comment by ldez** [comment]
Hello,

I change some elements related to facts (they are used by staticcheck) maybe the problem will be fixed with the next release (v1.50.0)

**Comment by fho** [comment]
The issue happend again with golangci-lint 1.50.0.

**Comment by nocive** [comment]
bump

**Comment by iwata** [comment]
Happened with golangci-lint v1.54.2, too.

**Comment by EvgenySerg** [comment]
Have this issue also when dealing with switch/case:
nolint:exhaustive` is unused for linter "exhaustive" (nolintlint)
 v1.55.2

**Comment by ldez** [comment]
If you are facing the same problem, the best way to contribute is to provide the following information:

```
golangci-lint --version
```

```
cat .golangci.yml
```

```
go version && go env
```

```
golangci-lint cache clean
golangci-lint run -v
```

And a code example or link to a public repository.

**Comment by rliebz** [comment]
Here's a reliable repro:

`main.go`:
```go

package main

import (
	"fmt"

	"github.com/jackc/pgx/v5"
)

func main() {
	var conn *pgx.Conn
	fmt.Println(conn.PgConn().CheckConn() == nil)
	fmt.Println(conn.PgConn().CheckConn() == nil) //nolint: staticcheck
}
```

`go.mod`:
```go
module example.com/nolintrepro

go 1.22.0

require github.com/jackc/pgx/v5 v5.5.3

require (
	github.com/jackc/pgpassfile v1.0.0 // indirect
	github.com/jackc/pgservicefile v0.0.0-20221227161230-091c0ba34f0a // indirect
	golang.org/x/crypto v0.17.0 // indirect
	golang.org/x/text v0.14.0 // indirect
)
```

`.golangci.yml`:
```
linters:
  disable-all: true
  enable:
    - staticcheck
    - nolintlint
```

```bash
$ go install github.com/golangci/golangci-lint/cmd/golangci-lint@latest
$ golangci-lint --version
golangci-lint has version v1.56.2 built with go1.22.0 from (unknown, mod sum: "h1:dgQzlWHgNbCqJjuxRJhFEnHDVrrjuTGQHJ3RIZMpp/o=") on (unknown)
```

```bash
$ go version && go env
go version go1.22.0 darwin/arm64
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/me/Library/Caches/go-build'
GOENV='/Users/me/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/me/.local/share/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/me/.local/share/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/opt/homebrew/Cellar/go/1.22.0/libexec'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/opt/homebrew/Cellar/go/1.22.0/libexec/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.22.0'
GCCGO='gccgo'
AR='ar'
CC='cc'
CXX='c++'
CGO_ENABLED='1'
GOMOD='/Users/me/nolintrepro/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/yd/g6bkd_fd7qs40y6q_8w2fh2w0000gn/T/go-build3765820227=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

```
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/me/nolintrepro /Users/me /Users /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 2 linters: [nolintlint staticcheck] 
INFO [loader] Go packages loading at mode 575 (deps|name|types_sizes|compiled_files|exports_file|files|imports) took 523.977875ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 161µs 
INFO [linters_context/goanalysis] analyzers took 2.663033296s with top 10 stages: buildir: 2.039481668s, fact_deprecated: 150.111285ms, nilness: 126.620753ms, fact_purity: 125.42288ms, SA5012: 115.692744ms, typedness: 103.558214ms, SA4006: 222µs, isgenerated: 130.292µs, SA4027: 99.542µs, SA9004: 88.75µs 
INFO [runner] Issues before processing: 3, after processing: 1 
INFO [runner] Processors filtering stat (out/in): path_shortener: 1/1, cgo: 3/3, filename_unadjuster: 3/3, uniq_by_line: 1/1, nolint: 1/3, severity-rules: 1/1, fixer: 1/1, skip_files: 3/3, autogenerated_exclude: 3/3, exclude: 3/3, max_from_linter: 1/1, skip_dirs: 3/3, identifier_marker: 3/3, diff: 1/1, max_same_issues: 1/1, source_code: 1/1, path_prefixer: 1/1, sort_results: 1/1, path_prettifier: 3/3, exclude-rules: 3/3, max_per_file_from_linter: 1/1 
INFO [runner] processing took 519.497µs with stages: exclude-rules: 189.292µs, nolint: 94.251µs, autogenerated_exclude: 71.416µs, identifier_marker: 69.291µs, source_code: 52.833µs, path_prettifier: 31.583µs, skip_dirs: 4.959µs, uniq_by_line: 1.417µs, cgo: 1.041µs, max_same_issues: 749ns, path_shortener: 458ns, max_from_linter: 417ns, filename_unadjuster: 292ns, sort_results: 292ns, diff: 209ns, fixer: 208ns, max_per_file_from_linter: 208ns, skip_files: 167ns, severity-rules: 166ns, exclude: 166ns, path_prefixer: 82ns 
INFO [runner] linters took 1.05964275s with stages: goanalysis_metalinter: 1.059068917s 
main.go:11:14: SA1019: conn.PgConn().CheckConn is deprecated: CheckConn is deprecated in favor of Ping. CheckConn cannot detect all types of broken connections where the write would still appear to succeed. Prefer Ping unless on a high latency connection. (staticcheck)
	fmt.Println(conn.PgConn().CheckConn() == nil)
	           ^
INFO File cache stats: 1 entries of total size 213B 
INFO Memory: 19 samples, avg is 131.1MB, max is 309.6MB 
INFO Execution took 1.774828083s          
```

This is the correct behavior—`staticcheck` is flagging the line without the `nolint` directive.

But if we modify the file (which I assume affects the cache), we get the incorrect behavior. I added a newline after `var conn *pgx.Conn` to make a change that should be equivalent from an AST perspective. Then ran `golangci-lint run -v` again:

```
INFO [config_reader] Config search paths: [./ /Users/me/nolintrepro /Users/me /Users /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 2 linters: [nolintlint staticcheck] 
INFO [loader] Go packages loading at mode 575 (exports_file|files|imports|types_sizes|compiled_files|deps|name) took 531.612584ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 198.542µs 
INFO [linters_context/goanalysis] analyzers took 8.812332ms with top 10 stages: buildir: 5.229666ms, SA5012: 532.211µs, fact_purity: 447.049µs, nilness: 416.708µs, fact_deprecated: 405.157µs, typedness: 365.661µs, SA4001: 78.208µs, SA1021: 77.834µs, SA6005: 70.25µs, SA4026: 65.584µs 
INFO [runner] Processors filtering stat (out/in): exclude: 1/1, diff: 1/1, source_code: 1/1, fixer: 1/1, cgo: 1/1, filename_unadjuster: 1/1, skip_dirs: 1/1, autogenerated_exclude: 1/1, uniq_by_line: 1/1, severity-rules: 1/1, skip_files: 1/1, exclude-rules: 1/1, nolint: 1/1, max_same_issues: 1/1, max_from_linter: 1/1, path_shortener: 1/1, path_prettifier: 1/1, identifier_marker: 1/1, max_per_file_from_linter: 1/1, path_prefixer: 1/1, sort_results: 1/1 
INFO [runner] processing took 224.418µs with stages: nolint: 57.251µs, autogenerated_exclude: 51.125µs, source_code: 27.209µs, path_prettifier: 25.876µs, identifier_marker: 25.125µs, exclude-rules: 25.042µs, skip_dirs: 7.542µs, uniq_by_line: 1.042µs, max_same_issues: 1.042µs, max_from_linter: 584ns, cgo: 541ns, path_shortener: 459ns, filename_unadjuster: 292ns, fixer: 208ns, skip_files: 208ns, exclude: 208ns, severity-rules: 207ns, sort_results: 166ns, max_per_file_from_linter: 125ns, diff: 125ns, path_prefixer: 41ns 
INFO [runner] linters took 88.876792ms with stages: goanalysis_metalinter: 88.6125ms 
main.go:13:48: directive `//nolint: staticcheck` is unused for linter "staticcheck" (nolintlint)
	fmt.Println(conn.PgConn().CheckConn() == nil) //nolint: staticcheck
	                                             ^
INFO File cache stats: 1 entries of total size 214B 
INFO Memory: 10 samples, avg is 30.4MB, max is 51.4MB 
INFO Execution took 819.347ms                 
```

Now it complains about the unused `nolint` directive, and `staticcheck` doesn't complain at all. 

If you run `golangci-lint cache clean`, the correct behavior returns until the file is modified again.

**Comment by ldez** [comment]
Thank you for the repro.

As a workaround for `staticcheck`, I recommend using the following example instead of `nolint` directives:

```yml
issues:
  exclude-rules:
    - linters:
        - staticcheck
      text: "SA1019: conn.PgConn().CheckConn is deprecated: CheckConn is deprecated in favor of Ping"
```

Side note: a directive should not contain space after the `:`:
- `nolint: xxx` not valid
- `nolint:xxx` valid.

**Comment by stevenh** [comment]
Seeing this in combination with `ireturn` too, what's interesting is the error goes away for a time if you clean the cache with `golangci-lint cache clean`.

```shell
pkg/mything.go:9:78: directive `//nolint:ireturn // Returns one of multiple types which implement MyThinger.` is unused for linter "ireturn" (nolintlint)
func Testfunc() MyThinger { //nolint:ireturn // Returns one of multiple types which implement MyThinger.
```

**Comment by ldez** [comment]
I think this issue is the same as #4423, I will check my fix #4424 on that.

**Comment by ZackaryWelch** [comment]
I've just started noticing this on 1.63.4

**Comment by magnetikonline** [comment]
> I've just started noticing this on 1.63.4

Likewise @ZackaryWelch - just implementing `nolintlint` and seeing this.

**Comment by ldez** [comment]
If you are facing the same problem, the best way to contribute is to provide the following information:

```
golangci-lint --version
```

```
cat .golangci.yml
```

```
go version && go env
```

```
golangci-lint cache clean
golangci-lint run -v
```

**And a code example or link to a public repository.**

**Comment by bwplotka** [comment]
We are hitting this in Prometheus too, with GH action golangci-lint: https://github.com/prometheus/prometheus/issues/16203

It's flaky, 50% chances it brings false positive. Rerunning often helps.

We are disabling `nolintlint` fo now https://github.com/prometheus/prometheus/pull/16204 as we see contributors trying to remove the `nolint` directive to resolve false positive only to get a failure again from revive 🙈 

### Example case

We are hitting this on GH actions for 1.64.6 golangci-lint: https://github.com/prometheus/prometheus/actions/runs/13794350636/job/38582148069

* CI config: https://github.com/prometheus/prometheus/blob/30d04792caee71631f9d250f64cfc08e4c61b935/.github/workflows/ci.yml#L191
* golangci.yaml (before disabling nolintlint): https://github.com/prometheus/prometheus/blob/30d04792caee71631f9d250f64cfc08e4c61b935/.golangci.yml#L7 

Pasting logs in case of the GH action logs will expire:

```
Checking for go.mod: go.mod
  Cache hit for: golangci-lint.cache-Linux-2879-2c745dd6972f078a7e744b34011c2d4048562193
  Received 32[15](https://github.com/prometheus/prometheus/actions/runs/13794350636/job/38582148069#step:5:16)192 of 3215192 (100.0%), 9.0 MBs/sec
  Cache Size: ~3 MB (3215192 B)
  /usr/bin/tar -xf /home/runner/work/_temp/de86874b-1a55-4cf6-b9cd-94a93a3d5f95/cache.tzst -P -C /home/runner/work/prometheus/prometheus --use-compress-program unzstd
  Cache restored successfully
  Restored cache for golangci-lint from key 'golangci-lint.cache-Linux-2879-2c745dd6972f078a7e744b34011c2d4048562193' in 1445ms
  Finding needed golangci-lint version...
  Installation mode: binary
  Installing golangci-lint binary v1.64.6...
  Downloading binary https://github.com/golangci/golangci-lint/releases/download/v1.64.6/golangci-lint-1.64.6-linux-amd64.tar.gz ...
  /usr/bin/tar xz --overwrite --warning=no-unknown-keyword --overwrite -C /home/runner -f /home/runner/work/_temp/d2246a6e-8b43-457f-a8f5-cd4b[16](https://github.com/prometheus/prometheus/actions/runs/13794350636/job/38582148069#step:5:17)aa8f71
  Installed golangci-lint into /home/runner/golangci-lint-1.64.6-linux-amd64/golangci-lint in 463ms
  Prepared env in 1911ms
run golangci-lint
  Running [/home/runner/golangci-lint-1.64.6-linux-amd64/golangci-lint config path] in [/home/runner/work/prometheus/prometheus] ...
  Running [/home/runner/golangci-lint-1.64.6-linux-amd64/golangci-lint config verify] in [/home/runner/work/prometheus/prometheus] ...
  Running [/home/runner/golangci-lint-1.64.6-linux-amd64/golangci-lint run  --verbose] in [/home/runner/work/prometheus/prometheus] ...
  Error: tsdb/ooo_head.go:75:1: directive `//nolint:revive` is unused for linter "revive" (nolintlint)
  //nolint:revive
  ^
  
  level=info msg="golangci-lint has version 1.64.6 built with go1.24.0 from d574f356 on 2025-03-02T23:27:45Z"
  level=info msg="[config_reader] Config search paths: [./ /home/runner/work/prometheus/prometheus /home/runner/work/prometheus /home/runner/work /home/runner /home /]"
  level=info msg="[config_reader] Used config file .golangci.yml"
  level=info msg="[goenv] Read go env for 3.628914ms: map[string]string{\"GOCACHE\":\"/home/runner/.cache/go-build\", \"GOROOT\":\"/opt/hostedtoolcache/go/1.24.1/x64\"}"
  level=info msg="[lintersdb] Active 25 linters: [depguard errcheck errorlint exptostd gocritic godot gofumpt goimports gosimple govet ineffassign loggercheck misspell nilnesserr nolintlint perfsprint predeclared revive sloglint staticcheck testifylint unconvert unused usestdlibvars whitespace]"
  level=info msg="[loader] Go packages loading at mode 8767 (deps|files|imports|name|types_sizes|compiled_files|exports_file) took 7.432398889s"
  level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 182.425377ms"
  level=info msg="[linters_context/goanalysis] analyzers took 1m46.207040965s with top 10 stages: buildir: 13.192511297s, goimports: 12.95784788s, gofumpt: 8.426084967s, the_only_name: 8.275105025s, S1038: 5.64723[17](https://github.com/prometheus/prometheus/actions/runs/13794350636/job/38582148069#step:5:18)73s, gocritic: 5.628474353s, unused: 5.034059602s, whitespace: 4.393366523s, unconvert: 3.816960858s, buildssa: 3.486442007s"
  level=info msg="[runner/skip_dirs] Skipped 24 issues from dir storage/remote/otlptranslator/prometheusremotewrite by pattern storage/remote/otlptranslator/prometheusremotewrite"
  level=info msg="[runner/skip_dirs] Skipped 17 issues from dir storage/remote/otlptranslator/prometheus by pattern storage/remote/otlptranslator/prometheus"
  level=info msg="[runner/skip_dirs] Skipped 6 issues from dir documentation/examples/custom-sd/adapter by pattern (^|/)examples($|/)"
  level=info msg="[runner/skip_dirs] Skipped 3 issues from dir documentation/examples/custom-sd/adapter-usage by pattern (^|/)examples($|/)"
  level=info msg="[runner/exclusion_rules] Skipped 20 issues by rules: [Text: \"(?i)stdmethods: method Seek.* should have signature Seek\", Linters: \"govet\"]"
  level=info msg="[runner/exclusion_rules] Skipped 569 issues by rules: [Text: \"(?i)exported (.+) should have comment( \\\\(or a comment on this block\\\\))? or be unexported\", Linters: \"revive\"]"
  level=info msg="[runner/exclusion_rules] Skipped 247 issues by rules: [Path: \"_test.go\", Linters: \"errcheck\"]"
  level=info msg="[runner/exclusion_rules] Skipped 2 issues by rules: [Path: \"tsdb/head_wal.go\", Linters: \"errorlint\"]"
  level=info msg="[runner/exclusion_rules] Skipped 52 issues by rules: [Text: \"(?i)fmt.Sprintf can be replaced with string concatenation\", Linters: \"perfsprint\"]"
  level=info msg="[runner/exclusion_rules] Skipped 356 issues by rules: [Text: \"(?i)Error return value of .((os\\\\.)?std(out|err)\\\\..*|.*Close|.*Flush|os\\\\.Remove(All)?|.*print(f|ln)?|os\\\\.(Un)?Setenv). is not checked\", Linters: \"errcheck\"]"
  level=info msg="[runner/exclusion_rules] Skipped 9 issues by rules: [Text: \"(?i)appendAssign\", Linters: \"gocritic\"]"
  level=info msg="[runner/exclusion_rules] Skipped 72 issues by rules: [Source: \"(?i)^// ===\", Linters: \"godot\"]"
  level=info msg="[runner] Issues before processing: 1977, after processing: 1"
  level=info msg="[runner] Processors filtering stat (in/out): invalid_issue: 1977/1977, exclusion_paths: 1977/1977, skip_files: 1977/1534, generated_file_filter: 1484/1414, diff: 1/1, uniq_by_line: 1/1, max_per_file_from_linter: 1/1, max_same_issues: 1/1, filename_unadjuster: 1977/1977, skip_dirs: 1534/1484, identifier_marker: 1414/1414, nolint_filter: 87/1, fixer: 1/1, max_from_linter: 1/1, source_code: 1/1, severity-rules: 1/1, cgo: 1977/1977, path_relativity: 1977/1977, exclusion_rules: 1414/87, path_shortener: 1/1, sort_results: 1/1, path_absoluter: 1977/1977, path_prettifier: 1/1"
  level=info msg="[runner] processing took 54.838729ms with stages: nolint_filter: 35.67097ms, exclusion_rules: 9.666635ms, generated_file_filter: 5.331944ms, path_relativity: 1.56576ms, skip_files: 1.207978ms, skip_dirs: 804.571µs, invalid_issue: 160.491µs, cgo: 127.4µs, path_absoluter: 113.413µs, identifier_marker: 84.528µs, filename_unadjuster: 67.266µs, source_code: [18](https://github.com/prometheus/prometheus/actions/runs/13794350636/job/38582148069#step:5:19).835µs, path_shortener: 12.815µs, uniq_by_line: 1.714µs, sort_results: 1.353µs, max_same_issues: 951ns, exclusion_paths: 641ns, fixer: 351ns, diff: 340ns, max_per_file_from_linter: 320ns, path_prettifier: 211ns, max_from_linter: 142ns, severity-rules: 100ns"
  level=info msg="[runner] linters took 6.95610147s with stages: goanalysis_metalinter: 6.90113899s"
  level=info msg="File cache stats: 2 entries of total size 73.4KiB"
  level=info msg="Memory: 136 samples, avg is 418.2MB, max is 1467.3MB"
  level=info msg="Execution took 14.574[19](https://github.com/prometheus/prometheus/actions/runs/13794350636/job/38582148069#step:5:20)4377s"
  
  Error: issues found
  Ran golangci-lint in 14696ms
```

This failed for this [code](https://github.com/prometheus/prometheus/blob/30d04792caee71631f9d250f64cfc08e4c61b935/tsdb/ooo_head.go#L75): 

```
// ToEncodedChunks returns chunks with the samples in the OOOChunk.
//
//nolint:revive
func (o *OOOChunk) ToEncodedChunks(mint, maxt int64) (chks []memChunk, err error) {
```

Removing `nolint` entry fails with that unexported-return revive error as expected.

It happens in various places, likely when someone is touching the same directory in a PR 🤔

**Comment by dehaansa** [comment]
Also observing this in OpenTelemetry, inconsistent failures on a `//nolint:staticcheck` for a deprecated field.

Recent run with failure: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/14279045422/job/40026274454?pr=39029

**Comment by ldez** [comment]
I used Prometheus as a reproducible example.

The problem is not related to `nolinlinit` itself:

<details>
<summary>OK</summary>

```
INFO [runner] Issues before processing: 1956, after processing: 0
INFO [runner] Processors filtering stat (in/out): exclusion_paths: 1956/1511, exclusion_rules: 1441/100, nolint_filter: 100/0, generated_file_filter: 1511/1441
```

</details> 


<details>
<summary>KO</summary>

```
INFO [runner] Issues before processing: 1953, after processing: 3
INFO [runner] Processors filtering stat (in/out): exclusion_paths: 1953/1508, exclusion_rules: 1438/97, nolint_filter: 97/3, generated_file_filter: 1508/1438
```

</details> 


As you can see, 3 reports are missing (1956 vs 1953), so `nolinlinit` is right, but the problem is related to reports not being created.

I need to investigate more.

**Comment by ldez** [comment]
I'm digging into commits, issues, and PRs that have been linked to this issue.

---

`prometheus/prometheus`

The problem is related to revive.
I opened PRs on revive to fix the problem.

And I will open a PR on `prometheus/prometheus` when the fix is merged and released.

---

`open-telemetry/opentelemetry-collector-contrib`

I run a dozen analyses on the module `exporter/datadogexporter`, for now, I'm not reproducing the problem.
But this repo contains 320 modules (!!!), so it can be complex to reduce the exact context.

---

The problem from the comment https://github.com/golangci/golangci-lint/issues/3228#issuecomment-1957676891 is related to leading spaces inside the nolint directives.

---

`argoproj/argo-cd`

The problem is also related to leading spaces inside the nolint directives.
- https://github.com/argoproj/argo-cd/blob/e4a8c9ee980b150167b37afa27385d6ae564f19d/util/argo/argo.go#L1007
- https://github.com/argoproj/argo-cd/blob/e4a8c9ee980b150167b37afa27385d6ae564f19d/util/lua/lua.go#L191
- https://github.com/argoproj/argo-cd/blob/e4a8c9ee980b150167b37afa27385d6ae564f19d/reposerver/repository/repository.go#L923

I will open a PR on argo-cd to fix that.

---

TODO:
- [ ] `opendatahub-io/opendatahub-operator`
- [ ] `shiftstack/cluster-api-provider-openstack`
- [ ] `cerbos/cerbos`
- [ ] `sourcegraph/sourcegraph-public-snapshot`

**Comment by rliebz** [comment]
> The problem from the comment [#3228 (comment)](https://github.com/golangci/golangci-lint/issues/3228#issuecomment-1957676891) is related to leading spaces inside the nolint directives.

Thanks for looking into this—unfortunately I'm still getting a reproduction on 2.0.2 with the leading space removed. 


---

Updated version of the original repro with golangci-lint 2.0.2 and the spacing on the nolint directive fixed:

<details>

```sh
#!/bin/sh -e

## setup module
rm -rf nolintrepro
mkdir nolintrepro && cd $_

## go.mod
cat > go.mod <<EOF
module example.com/nolintrepro

go 1.21

require github.com/jackc/pgx/v5 v5.5.3

require (
	github.com/jackc/pgpassfile v1.0.0 // indirect
	github.com/jackc/pgservicefile v0.0.0-20221227161230-091c0ba34f0a // indirect
	golang.org/x/crypto v0.17.0 // indirect
	golang.org/x/text v0.14.0 // indirect
)
EOF

## main.go
cat > main.go <<EOF
package main

import (
	"fmt"

	"github.com/jackc/pgx/v5"
)

func main() {
	var conn *pgx.Conn
	fmt.Println(conn.PgConn().CheckConn() == nil)
	fmt.Println(conn.PgConn().CheckConn() == nil) //nolint:staticcheck
}
EOF

go mod tidy

## .golangci.yml
cat > .golangci.yml <<EOF
version: "2"
linters:
  default: none
  enable:
    - nolintlint
    - staticcheck
EOF

## Run
golangci-lint run

## Clean
cd ..
rm -rf nolintrepro
```

</details>

<details><summary>version and env</summary>

```console
$ golangci-lint --version
golangci-lint has version 2.0.2 built with go1.24.1 from 2b224c2c on 2025-03-25T21:36:18Z
```

```console
$ go version && go env -changed
go version go1.24.2 darwin/arm64
```

</details>


<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO golangci-lint has version 2.0.2 built with go1.24.1 from 2b224c2c on 2025-03-25T21:36:18Z
INFO [config_reader] Config search paths: [./ /Users/me/Sandbox/nolintrepro /Users/me/Sandbox /Users/me /Users /]
INFO [config_reader] Used config file .golangci.yml
INFO maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
INFO [goenv] Read go env for 14.23125ms: map[string]string{"GOCACHE":"/Users/me/Library/Caches/go-build", "GOROOT":"/opt/homebrew/Cellar/go/1.24.2/libexec"}
INFO [lintersdb] Active 2 linters: [nolintlint staticcheck]
INFO [loader] Go packages loading at mode 8767 (compiled_files|files|name|types_sizes|deps|exports_file|imports) took 1.186928625s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 251.875µs
INFO [linters_context/goanalysis] analyzers took 3.7664686s with top 10 stages: buildir: 3.109269169s, fact_purity: 144.287161ms, nilness: 134.310297ms, SA5012: 129.473035ms, fact_deprecated: 112.895712ms, typedness: 108.396918ms, tokenfileanalyzer: 24.800508ms, SA1019: 222.917µs, SA6006: 222.792µs, SA4001: 198.542µs
INFO [runner] Issues before processing: 3, after processing: 1
INFO [runner] Processors filtering stat (in/out): path_relativity: 3/3, exclusion_rules: 3/3, severity-rules: 1/1, exclusion_paths: 3/3, generated_file_filter: 3/3, diff: 1/1, max_per_file_from_linter: 1/1, max_same_issues: 1/1, path_shortener: 1/1, filename_unadjuster: 3/3, invalid_issue: 3/3, max_from_linter: 1/1, path_prettifier: 1/1, nolint_filter: 3/1, fixer: 1/1, uniq_by_line: 1/1, source_code: 1/1, sort_results: 1/1, path_absoluter: 3/3, cgo: 3/3
INFO [runner] processing took 135.378µs with stages: nolint_filter: 51.084µs, generated_file_filter: 32.791µs, source_code: 21.958µs, cgo: 16µs, sort_results: 4.75µs, path_relativity: 2.458µs, uniq_by_line: 1.251µs, max_same_issues: 959ns, max_per_file_from_linter: 917ns, invalid_issue: 875ns, path_shortener: 375ns, max_from_linter: 375ns, path_absoluter: 292ns, fixer: 292ns, filename_unadjuster: 250ns, path_prettifier: 250ns, exclusion_paths: 250ns, severity-rules: 125ns, diff: 84ns, exclusion_rules: 42ns
INFO [runner] linters took 1.258803084s with stages: goanalysis_metalinter: 1.25861875s
main.go:11:14: SA1019: conn.PgConn().CheckConn is deprecated: CheckConn is deprecated in favor of Ping. CheckConn cannot detect all types of broken connections where the write would still appear to succeed. Prefer Ping unless on a high latency connection. (staticcheck)
        fmt.Println(conn.PgConn().CheckConn() == nil)
                    ^
1 issues:
* staticcheck: 1
INFO File cache stats: 1 entries of total size 212B
INFO Memory: 26 samples, avg is 158.7MB, max is 432.1MB
INFO Execution took 2.460576084s                                      
```

</details>

This is the correct behavior—`staticcheck` is flagging the line without the `nolint` directive.

But if we modify the file, we get the incorrect behavior. I added a newline after `var conn *pgx.Conn` to make a change that should be equivalent from an AST perspective. Then ran `golangci-lint run -v` again:


<details>

```
INFO golangci-lint has version 2.0.2 built with go1.24.1 from 2b224c2c on 2025-03-25T21:36:18Z
INFO [config_reader] Config search paths: [./ /Users/me/Sandbox/nolintrepro /Users/me/Sandbox /Users/me /Users /]
INFO [config_reader] Used config file .golangci.yml
INFO maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
INFO [goenv] Read go env for 16.550542ms: map[string]string{"GOCACHE":"/Users/me/Library/Caches/go-build", "GOROOT":"/opt/homebrew/Cellar/go/1.24.2/libexec"}
INFO [lintersdb] Active 2 linters: [nolintlint staticcheck]
INFO [loader] Go packages loading at mode 8767 (files|name|types_sizes|compiled_files|deps|exports_file|imports) took 178.570208ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 139µs
INFO [linters_context/goanalysis] analyzers took 0s with no stages
INFO [runner] Processors filtering stat (in/out): cgo: 1/1, filename_unadjuster: 1/1, diff: 1/1, max_from_linter: 1/1, severity-rules: 1/1, path_absoluter: 1/1, nolint_filter: 1/1, uniq_by_line: 1/1, max_same_issues: 1/1, sort_results: 1/1, path_relativity: 1/1, exclusion_paths: 1/1, exclusion_rules: 1/1, fixer: 1/1, max_per_file_from_linter: 1/1, path_prettifier: 1/1, invalid_issue: 1/1, generated_file_filter: 1/1, source_code: 1/1, path_shortener: 1/1
INFO [runner] processing took 114.25µs with stages: nolint_filter: 55µs, generated_file_filter: 26.834µs, source_code: 21.417µs, sort_results: 2.167µs, cgo: 1.792µs, path_relativity: 1.417µs, max_same_issues: 1.291µs, uniq_by_line: 1.124µs, invalid_issue: 542ns, path_shortener: 459ns, max_from_linter: 416ns, max_per_file_from_linter: 291ns, exclusion_paths: 250ns, fixer: 250ns, path_prettifier: 250ns, path_absoluter: 209ns, filename_unadjuster: 209ns, exclusion_rules: 166ns, diff: 84ns, severity-rules: 82ns
INFO [runner] linters took 42.807875ms with stages: goanalysis_metalinter: 42.646125ms
main.go:13:48: directive `//nolint:staticcheck` is unused for linter "staticcheck" (nolintlint)
        fmt.Println(conn.PgConn().CheckConn() == nil) //nolint:staticcheck
                                                      ^
1 issues:
* nolintlint: 1
INFO File cache stats: 1 entries of total size 213B
INFO Memory: 4 samples, avg is 40.8MB, max is 55.1MB
INFO Execution took 237.341542ms                                                
```

</details>

Now it complains about the unused `nolint` directive, and `staticcheck` doesn't complain at all.

If you run `golangci-lint cache clean`, the correct behavior returns until the file is modified again.


---

Based on your progress with `revive` I tried to see if I could reproduce flakiness with `staticcheck` alone, outside of `golangci-lint`. Using the same `main.go` file from above, the output is always the same:

```console
$ go run honnef.co/go/tools/cmd/staticcheck@latest .
main.go:11:14: conn.PgConn().CheckConn is deprecated: CheckConn is deprecated in favor of Ping. CheckConn cannot detect all types of broken connections where the write would still appear to succeed. Prefer Ping unless on a high latency connection.  (SA1019)
main.go:12:14: conn.PgConn().CheckConn is deprecated: CheckConn is deprecated in favor of Ping. CheckConn cannot detect all types of broken connections where the write would still appear to succeed. Prefer Ping unless on a high latency connection.  (SA1019)
exit status 1
```

Editing the file and re-running does not change the behavior of the `staticcheck` standalone program.

**Comment by ldez** [comment]
FYI, I found the root cause, but this is far from trivial to fix.

**Comment by gillesbouvier-qz** [comment]
Here is a possible snippet. Without the `//nolint:all`, it complains with a `ireturn` error on the last two functions. With it, it complains that it's unused.

```GO
// The slog.Handler interface requires returning the slog.Handler interface,
// applying a waiver at the package level, as this error keeps tripping the linter.
//
//nolint:all
package noophandler

import (
	"context"
	"log/slog"
)

// NoopHandler is a no-op handler that discards all log messages.
type NoopHandler struct{}

func (h *NoopHandler) Enabled(context.Context, slog.Level) bool  { return false }
func (h *NoopHandler) Handle(context.Context, slog.Record) error { return nil }

// The slog.Handler interface requires returning the slog.Handler interface
func (h *NoopHandler) WithAttrs([]slog.Attr) slog.Handler { return h }

// The slog.Handler interface requires returning the slog.Handler interface
func (h *NoopHandler) WithGroup(string) slog.Handler { return h }
```

**Comment by ldez** [comment]
<details><summary>I used the following files</summary>

```yml
version: "2"

linters:
  default: none
  enable:
    - ireturn
  settings:
    ireturn:
      reject:
        - anon
        - empty
        - stdlib
```

```go
package noophandler

import (
	"context"
	"log/slog"
)

// NoopHandler is a no-op handler that discards all log messages.
type NoopHandler struct{}

func (h *NoopHandler) Enabled(context.Context, slog.Level) bool  { return false }
func (h *NoopHandler) Handle(context.Context, slog.Record) error { return nil }

// WithAttrs ...
//nolint:ireturn
func (h *NoopHandler) WithAttrs([]slog.Attr) slog.Handler { return h }

// WithGroup ...
//nolint:ireturn
func (h *NoopHandler) WithGroup(string) slog.Handler { return h }
```

</details>


I tried to create the problem in several ways, but everything works as expected.
- `nolint` directive works
- no unused reports

**Comment by vtopc** [comment]
The same for `revive`:
```
directive `//nolint:revive // reason` is unused for linter "revive" (nolintlint)
```
when trying to silent a blank import(`blank-imports`).
using `golangci-lint has version 2.4.0 built with go1.25.0 from 43d03392 on 2025-08-13T23:36:29Z`

**Comment by Gchbg** [comment]
I'm encountering this quite reliably, please let me know if I can help investigate.

**Comment by RainbowMango** [comment]
> I'm encountering this quite reliably, please let me know if I can help investigate.

Kindly ping @ldez here

**Comment by jandubois** [comment]
I was seeing this issue locally today as well. It repeatedly showed the (incorrect) warning.

After I ran `go tool golangci-lint cache clean` the warning went away.

**Comment by rostislaved** [comment]
Happened with golangci-lint v2.11.3

**Comment by rostislaved** [comment]
I can reproduce what looks like the same underlying `nolintlint` flakiness with `revive`, not only with `staticcheck`.

  Environment:
  - `golangci-lint v2.11.3`

  Minimal repro:

  `.golangci.yml`
  ```yaml
  version: "2"

  linters:
    default: none
    enable:
      - revive
      - nolintlint
    settings:
      revive:
        enable-all-rules: true
        rules:
          - name: add-constant
            arguments:
              - allow-ints: "1,2"
          - name: enforce-switch-style
            arguments: ["allowNoDefault"]
          - name: package-comments
            disabled: true
  ```
` main.go`
  ```
  package main

  import "awesomeProject10/x"

  //nolint:revive // identical-switch-branches
  func main() {
        switch 2 {
        case 1:
                _ = x.V
        case 2:
                _ = x.V
        }
  }
  ```
`  x/x.go`
  ```
  package x

  // V exported
  const V = 1
```

  Then repeatedly run:
```
  for ((i=1; i<=20; i++)); do
      echo "iteration: $i"

      golangci-lint cache clean
      rc=$?
      if ((rc != 0)); then
          echo "cache clean failed with exit code $rc"
          break
      fi

      golangci-lint run --fix
      rc=$?
      if ((rc != 0)); then
          echo "golangci-lint failed with exit code $rc"
          break
      fi
  done
```

  Observed behavior:

  - some iteration removes the valid //nolint:revive comment
  - after that, the next run reports:
    identical-switch-branches: "switch" with identical branches
  - if I restore the //nolint:revive comment manually, the cycle repeats

  Important details:
  - the issue disappears if nolintlint is disabled
  - the issue disappears if revive.enable-all-rules: true is removed
  - the issue disappears if the reference to x.V is replaced with a local symbol or a stdlib symbol
  - function calls / method chains / returned values are not required, just referencing an exported symbol from another local
    package is enough

**Comment by silverwind** [comment]
Here's another reproduction, fails deterministically:

https://github.com/silverwind/golangci-bug-3228

```go
var _ = ioutil.ReadAll   //nolint:staticcheck
var _ = ioutil.ReadFile  //nolint:staticcheck
var _ = ioutil.WriteFile //nolint:staticcheck
var _ = ioutil.TempDir   //nolint:staticcheck
```

Expected: nolintlint x4, staticcheck x1
Actual: nolintlint x3, staticcheck x1

If this is not the same bug, I can open a new issue.

**Comment by bombsimon** [comment]
@silverwind This is just `--max-same-issues` with its default value `3`.

```sh
› golangci-lint run
[...]
4 issues:
* nolintlint: 3
* staticcheck: 1
```

```sh
› golangci-lint run --max-same-issues 0
[...]
5 issues:
* nolintlint: 4
* staticcheck: 1
```

**Comment by silverwind** [comment]
Ah I see, that bad default always catches me (and my agent) off-guard. I will make a more robust repro. I had multiple variants fail in that repo before, all non-deterministic, so likely the real repro will also be.

Edit: Unable to reproduce my exact case, and it seems on one is interested in me assisting further on this, so I'm stopping.
